File: Kordia_backend/app/api/routes/playlists.py
from fastapi import APIRouter, HTTPException, Depends
import aiosqlite
from typing import List, Dict, Any
from app.database.repositories.playlists import PlaylistRepository
from app.dependencies import get_db

router = APIRouter(prefix="/playlists", tags=["playlists"])

@router.get("/")
async def get_playlists(db: aiosqlite.Connection = Depends(get_db)):
    repo = PlaylistRepository(db)
    return await repo.get_all_playlists()

@router.post("/")
async def create_playlist(data: Dict[str, Any], db: aiosqlite.Connection = Depends(get_db)):
    name = data.get("name")
    playlist_id = data.get("id")
    if not name or not playlist_id:
        raise HTTPException(status_code=400, detail="name and id required")
    repo = PlaylistRepository(db)
    return await repo.create_playlist(playlist_id, name)

@router.get("/{playlist_id}")
async def get_playlist(playlist_id: str, db: aiosqlite.Connection = Depends(get_db)):
    repo = PlaylistRepository(db)
    playlist = await repo.get_playlist(playlist_id)
    if not playlist:
        raise HTTPException(status_code=404, detail="Playlist not found")
    return playlist

@router.put("/{playlist_id}")
async def rename_playlist(playlist_id: str, data: Dict[str, Any], db: aiosqlite.Connection = Depends(get_db)):
    name = data.get("name")
    if not name:
        raise HTTPException(status_code=400, detail="name required")
    repo = PlaylistRepository(db)
    success = await repo.update_playlist_name(playlist_id, name)
    if not success:
        raise HTTPException(status_code=404, detail="Playlist not found")
    return {"status": "success"}

@router.delete("/{playlist_id}")
async def delete_playlist(playlist_id: str, db: aiosqlite.Connection = Depends(get_db)):
    repo = PlaylistRepository(db)
    success = await repo.delete_playlist(playlist_id)
    if not success:
        raise HTTPException(status_code=404, detail="Playlist not found")
    return {"status": "success"}

@router.post("/{playlist_id}/songs")
async def add_song(playlist_id: str, song: Dict[str, Any], db: aiosqlite.Connection = Depends(get_db)):
    repo = PlaylistRepository(db)
    if 'ytid' not in song or 'title' not in song:
        raise HTTPException(status_code=400, detail="Song needs ytid and title")
        
    if not await repo.get_playlist(playlist_id):
        raise HTTPException(status_code=404, detail="Playlist not found")
        
    await repo.add_song(playlist_id, song)
    # Refresh and return
    return await repo.get_playlist(playlist_id)

@router.delete("/{playlist_id}/songs/{ytid}")
async def remove_song(playlist_id: str, ytid: str, db: aiosqlite.Connection = Depends(get_db)):
    repo = PlaylistRepository(db)
    success = await repo.remove_song(playlist_id, ytid)
    if not success:
        raise HTTPException(status_code=404, detail="Song or Playlist not found")
    # Return updated playlist
    return await repo.get_playlist(playlist_id)

File: Kordia_backend/src/routes/playlists.ts
import type { FastifyInstance } from 'fastify';
import { getDb } from '../database/connection.js';
import { PlaylistRepository } from '../database/playlists.js';

export async function playlistRoutes(app: FastifyInstance): Promise<void> {
  app.get('/playlists', async () => {
    return new PlaylistRepository(getDb()).getAllPlaylists();
  });

  app.post('/playlists', {
    schema: {
      body: {
        type: 'object',
        required: ['id', 'name'],
        properties: { id: { type: 'string' }, name: { type: 'string' } },
      },
    },
  }, async (req, reply) => {
    const { id, name } = req.body as { id: string; name: string };
    return new PlaylistRepository(getDb()).createPlaylist(id, name);
  });

  app.get('/playlists/:playlistId', async (req, reply) => {
    const { playlistId } = req.params as { playlistId: string };
    const pl = new PlaylistRepository(getDb()).getPlaylist(playlistId);
    if (!pl) return reply.code(404).send({ error: 'Playlist not found' });
    return pl;
  });

  app.put('/playlists/:playlistId', {
    schema: {
      body: {
        type: 'object',
        required: ['name'],
        properties: { name: { type: 'string' } },
      },
    },
  }, async (req, reply) => {
    const { playlistId } = req.params as { playlistId: string };
    const { name } = req.body as { name: string };
    const ok = new PlaylistRepository(getDb()).updatePlaylistName(playlistId, name);
    if (!ok) return reply.code(404).send({ error: 'Playlist not found' });
    return { status: 'success' };
  });

  app.delete('/playlists/:playlistId', async (req, reply) => {
    const { playlistId } = req.params as { playlistId: string };
    const ok = new PlaylistRepository(getDb()).deletePlaylist(playlistId);
    if (!ok) return reply.code(404).send({ error: 'Playlist not found' });
    return { status: 'success' };
  });

  app.post('/playlists/:playlistId/songs', {
    schema: {
      body: {
        type: 'object',
        required: ['ytid', 'title'],
        properties: {
          ytid: { type: 'string' },
          title: { type: 'string' },
          artist: { type: 'string' },
          thumbnail: { type: 'string' },
        },
      },
    },
  }, async (req, reply) => {
    const { playlistId } = req.params as { playlistId: string };
    const song = req.body as { ytid: string; title: string; artist?: string; thumbnail?: string };
    const repo = new PlaylistRepository(getDb());
    if (!repo.getPlaylist(playlistId)) return reply.code(404).send({ error: 'Playlist not found' });
    repo.addSong(playlistId, song);
    return repo.getPlaylist(playlistId);
  });

  app.delete('/playlists/:playlistId/songs/:ytid', async (req, reply) => {
    const { playlistId, ytid } = req.params as { playlistId: string; ytid: string };
    const repo = new PlaylistRepository(getDb());
    const ok = repo.removeSong(playlistId, ytid);
    if (!ok) return reply.code(404).send({ error: 'Song or Playlist not found' });
    return repo.getPlaylist(playlistId);
  });
}
